### ISSUE TITLE
FR: Page Browser Preload Thumbnails

### ISSUE THREAD

**Issue by ryanwwest** [koreader/koreader#10811]
In page browser UI, a certain grid of MxN previous and next 'page thumbnails' of a PDF or ePUB are shown relative to the current reading page, and by swiping right or left (or up or down), the user can reveal the previous or next set of pages' preview thumbnails. This is very helpful for quick navigation to identify, e.g., section headings.

**Problem:** 90% of the time when I open Page browser, I intend to swipe left or right, and there is some delay that occurs as the preview thumbnails of each next/previous page are generated after swiping. On my Boox Max Lumi, a 3x3 grid of PDF pages took ~5.1 seconds to completely load. Once a set of pages is rendered, it seems to be cached as future swiping between set of pages is near instantaneous. 

**FR:** It would be nice to _pre-load_ the previous and next MxN grids of pages that are to be shown when the user swipes on Page browser. A balance must exist between caching for performance and memory usage, so I propose only preloading 1 MxN grid immediately before and 1 grid immediately after the currently-viewed grid of pages, and only start these pre-loading of thumbnails after the currently-viewing MxN grid is finished loading. If the user swipes and reveals one of these (potentially partially) pre-loaded grids, then once that grid is completely loaded, start preloading the next 1-distance neighbor MxN grids around that cell after.

I hope this would improve user-perceived load times, improving the usefulness of Page browser with hopefully negligible performance impact. Apologies if this has been done already—there has been a lot of work done on page browser lately but I didn't see anything specific to this.

**Comment by poire-z** [comment]
Not sure I'd want that by default:
- it's some cpu/battery wasted if you don't care for the prev and/or next pages (and the thumbnail cache may be trashed more often as some thumbnails are added that you may not even look at).
- navigation while everything seems drawn and calm may be sluggish while 10-30 thumbnails are being generated in the background
- I sometime get the "not yet cache" display/lag as a hint of my reading progress or pagebrowser navigation initial position or last time I launch it (ie. _wow, 3 panels back not cached = 12 pages I have read without launching Page browser to see when this boring chapter ends ! must have been not boring after all_ :))

Anyway, if you want to have it, you can drop this [user patch](https://github.com/koreader/koreader/wiki/User-patches):
[2-pageturbobrowser.lua.txt](https://github.com/koreader/koreader/files/12330672/2-pageturbobrowser.lua.txt)

```lua
local PageBrowserWidget = require("ui/widget/pagebrowserwidget")

local PageBrowserWidget_update_orig = PageBrowserWidget.update
PageBrowserWidget.update = function(self)
    -- Call original method
    PageBrowserWidget_update_orig(self)
    -- We're there with the page painted - and possibly some thumbnails
    -- not yet there and being generated and going to be updated.
    -- self.ui.thumbnail takes care of serializing tile requests, and
    -- cancelling scheduled ones when new requests come.
    -- So, we can just launch many for the next and previous PageBrowser
    -- views.
    -- Pre-generate the thumbnails for the next page
    local next_grid_page_start = self.focus_page - self.focus_page_shift + self.nb_grid_items
    local next_grid_page_end = next_grid_page_start + self.nb_grid_items - 1
    for idx=1, self.nb_grid_items do
        local p = next_grid_page_start + idx - 1
        if p >= 1 or p < self.nb_pages then
            -- We provide a dummy callback as we don't care about the tile
            self.ui.thumbnail:getPageThumbnail(p, self.grid_item_width, self.grid_item_height, self.requests_batch_id, function() end)
        end
    end
    -- Pre-generate the thumbnails for the prev page
    local prev_grid_page_start = self.focus_page - self.focus_page_shift - self.nb_grid_items
    local prev_grid_page_end = prev_grid_page_start + self.nb_grid_items - 1
    for idx=self.nb_grid_items, 1, -1 do
        local p = prev_grid_page_start + idx - 1
        if p >= 1 or p < self.nb_pages then
            self.ui.thumbnail:getPageThumbnail(p, self.grid_item_width, self.grid_item_height, self.requests_batch_id, function() end)
        end
    end
end
```

Just tested on the emulator - now idea how sluggish it can feel on a device.
If it really feels like a good thing, I guess it could be made a checkbox in the Pagebrowser hamburger menu.
(With a bit more cleverness, ie. generate alternatively one row for the next page, then one row for the prev page - so users navigating rows with swipe north/south get the next/prev row ready before the next/prev full pages.)

**Comment by ryanwwest** [comment]
Thank you. I'll try out this patch sometime. I guess if this were a company we'd do a user test to see how probable users are to swipe in page browser vs cpu hit.

> navigation while everything seems drawn and calm may be sluggish while 10-30 thumbnails are being generated in the background

I was thinking you'd pause generation of thumbnails whenever swiping (maybe finish the currently generating thumbnail at first then prioritize UI) - and isn't the only navigation action on this page swiping or tapping menu/close buttons? Maybe the single-threadedness makes this harder.

**Comment by ryanwwest** [comment]
Thanks again, @poire-z, I've been using this patch and it really makes a big difference with the responsiveness of the page browser without any noticeable ui lag so far. If you look at each thumbnail before swiping to the next set, there is zero lag now.

I'd still argue that the pros of this could outweigh the cons, especially after using it, but doesn't have to be that way. If you'd rather not consider that more, I'm fine to close this and point others at the user patch (maybe it should be linked on https://github.com/koreader/koreader/wiki/User-patches#available_patches?) if anyone else is as pedantic as me to find it.

**Comment by poire-z** [comment]
We can at least put this label ^.

Pinging another PageBrowserPowerUser @jonnyl2 for thoughts.

**Comment by jonnyl2** [comment]
I added the patch and will report back after some practical use.

**Comment by jonnyl2** [comment]
I didn't notice any problems and it makes the browsing experience noticeably faster. The patch came in really handy while doing a "Handmade ToC" (a feature still in development, also by @poire-z). While selecting a chapter start on one set of thumbnails, Page browser was already preparing the next set of thumbnails in the background. Excellent!

However, during normal reading I very rarely go forward in the Page browser. I normally only use it to preview the next few pages, or I skip entire chapters. So, for me personally, I'd welcome the feature as an option in the hamburger menu to be activated on demand.

**Comment by ryanwwest** [comment]
@jonnyl2 Good to hear. I'll have to look at the Handmade ToC - out of curiousity, does it add new ToC entries to the PDF file (assuming one doesn't exist) or just belong to the book Metadata?

Another thing I thought of is, if there is any battery-respecting setting (either user-enabled or one that monitors some OS signal to use less power), the this could be an example of something that gets disabled to save power.

**Comment by jonnyl2** [comment]
No, it doesn't touch any existing ToC but writes its own metadata. 
Details [here](https://github.com/koreader/koreader/issues/10193#issuecomment-1703927379).

Regarding battery saving options, there is an "Auto power save" mode, but AFAIK it currently only manages auto standby, suspend, and power off. Not sure if it could or would be adapted to handle additional controls.

**Comment by poire-z** [comment]
#10982 will bring this as an option in PageBrowser top left menu.
So, remove your user patch when updating to KOReader once it is merged.

**Comment by ryanwwest** [comment]
Thank you! I appreciate it being added in directly as an option.

**Comment by ryanwwest** [comment]
@poire-z I can make a FR for this to track, but noticed a separate Page Browser thing. In night mode, non-night-mode (day-mode?) thumbnails are still generated which is a little blinding when reading in the dark. Would that be a simple fix?

**Comment by poire-z** [comment]
^ Strange, I could have sweared I had tested it because I added:
https://github.com/koreader/koreader/blob/873503369c21f3ddc265c0623b477ceffd521f88/frontend/apps/reader/modules/readerthumbnail.lua#L434
(Unless I felt it was fine with daymode thumbnails at the time, in which case this was needed and the comment is true.)

Been trying to fix this for 30 mn as I thought it would indeed be simple - but not that simple :/
I'll keep that in my todo list, no need for opening another issue.

Do other nightmode users agree the thumbnails should be nightmodish ?
I also hope you'll be fine with this hatching on hidden flow pages:
![image](https://github.com/koreader/koreader/assets/24273478/55be0df9-b2f7-4ce3-97bf-64a339d2ca4c)

**Comment by ryanwwest** [comment]
Thanks, sounds great.

> I also hope you'll be fine with this hatching on hidden flow pages:

I actually don't understand the hidden flow pages so can't comment. Is there new documentation or an about section hidden in KOReader to explain them?

**Comment by jonnyl2** [comment]
>Do other nightmode users agree the thumbnails should be nightmodish ?

Yes, I agree they should be dark as well (so the thumbnails look like the pages in the Reader). I actually wanted to bring this up to you once but then I forgot about it, because I only very rarely use nightmode.

**Comment by jonnyl2** [comment]
>I actually don't understand the hidden flow pages so can't comment. Is there new documentation or an about section hidden buried in KOReader to explain them?

There's a help text in the Custum layout features menu option. The concept is simple (but the logic complex). You mark pages you don't want to include in your regular reading flow (e.g., first few pages and bibliography), and those are then skipped while reading. The page numbers and estimated remaining reading time will be adjusted.

**Comment by ryanwwest** [comment]
@jonnyl2 Thanks, I read them. I'm curious about how you use hidden flows. For me, it seems simple enough to just skip over the pages before e.g. Chapter 1 starts and then just keep reading—I know progress is a bit off though. But it seems like more work to mark the pages. For a book with a massive index at the end, I guess I could see very-inaccurate progress tracking being annoying.

Do you read applicable documents many times over for it to become worth it, or are there other benefits out of accurate progress/time-left indicators and not advancing past the conclusion I'm not seeing?

**Comment by poire-z** [comment]
> For me, it seems simple enough to just skip over the pages

I agree, and I've been doing that and I admit I'm still not using these new features :)
But there were various requests and different use cases that introducing this generic hidden flow stuff solve: see the list of closed issues at the bottom of first post of #10982 for some user stories.

**Comment by poire-z** [comment]
^ nightmode thumbnails when in nightmode done via #11091.
